R/mbbs.R

```diff
@@ -0,0 +1,362 @@
+#'
+#' @param ebird ebird dataset resulting from `get_ebird_data`
+#' @param taxonomy taxonomy dataset
+#' @param config mbbs configuration
+create_stop_level_counts_0 <- function(ebird, taxonomy, config = config) {
+  stop_ebird <- ebird |>
+    dplyr::filter(!is.na(stop_num)) |>
+    dplyr::select(
+      year, route, stop_num,
+      common_name, sci_name, count,
+      county, route_num
+    )
+
+  stop_xls <- get_stop_level_xls_data() |>
+    conform_taxonomy(taxonomy)
+
+  stop_obs <- process_obs_details(ebird) |>
+    select(year, route, common_name, sci_name, county, route_num, stop_data) |>
+    tidyr::unnest(cols = stop_data)
+
+  stop_transcribed <- get_stop_level_transcribed() |>
+    select(
+      year, route, stop_num,
+      common_name, sci_name, count,
+      county, route_num, source
+    )
+
+  logger::log_trace("Combining stop level data.")
+
+  dplyr::bind_rows(
+    stop_ebird |> mutate(source = "ebird"),
+    stop_xls |> mutate(source = "observer xls"),
+    stop_obs |> mutate(source = "obs details"),
+    stop_transcribed
+  ) |>
+    # Validity checks
+    (\(df) {
+      assertthat::assert_that(
+        !anyNA(df$stop_num),
+        msg = "Stop data has NA values in stop_num."
+      )
+
+      # Check that there are not duplicate entries with different counts
+      df |>
+        dplyr::group_by(year, route, stop_num, common_name) |>
+        dplyr::summarise(
+          n = dplyr::n(),
+          distinctcounts = n_distinct(count),
+          sources = paste(source, collapse = " & ")
+        ) |>
+        dplyr::filter(n > 1 & distinctcounts > 1) |>
+        (\(x) {
+          logger::log_error(
+            paste(
+              "Route {x$route}-{x$stop_num} has multiple entries",
+              "with differing counts",
+              "for {x$common_name}",
+              "in {x$year} coming from {x$sources}"
+            )
+          )
+        })()
+
+      # Log number of duplicate entries with the same count
+      df |>
+        dplyr::distinct(year, route, stop_num, common_name, count,
+          .keep_all = TRUE
+        ) |>
+        (\(x) {
+          logger::log_info(
+            paste(
+              "Stop-level data contains",
+              "{nrow(df) - nrow(x)} duplicated observations",
+              "(same count, different sources)"
+            )
+          )
+        })()
+
+      df
+    })()
+}
+
+#' Main function to create stop level count dataset
+#'
+#' @inheritParams create_stop_level_0
+create_stop_level_counts <- function(ebird, taxonomy, config = config) {
+  df <- create_stop_level_counts_0(ebird, taxonomy, config = config)
+
+  yrs_in <- dplyr::distinct(df, year, route, stop_num) |>
+    arrange(year, route, stop_num)
+
+  logger::log_trace("Preliminary stop-level data has {nrow(df)} observations.")
+
+  # remove duplicated observations
+  df <- df |>
+    dplyr::distinct(year, route, stop_num, common_name, count,
+      .keep_all = TRUE
+    )
```

Comment:
Causing parsing problems (chatham 4 2014 example). This is where the hueristic should be